dev/AnimatedIcon/AnimatedIcon.cpp

```diff
@@ -0,0 +1,521 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include "pch.h"
+#include "common.h"
+#include "AnimatedIcon.h"
+#include "RuntimeProfiler.h"
+#include "ResourceAccessor.h"
+#include <AnimatedVisuals\ProgressRingDeterminate.h>
+#include <AnimatedIconTestHooks.h>
+using namespace AnimatedVisuals;
+
+AnimatedIcon::AnimatedIcon()
+{
+    __RP_Marker_ClassById(RuntimeProfiler::ProfId_AnimatedIcon);
+    m_progressPropertySet = winrt::Window::Current().Compositor().CreatePropertySet();
+    m_progressPropertySet.InsertScalar(L"Progress", 0);
+
+    RegisterPropertyChangedCallback(winrt::IconElement::ForegroundProperty(), { this, &AnimatedIcon::OnForegroundPropertyChanged});
+}
+
+void AnimatedIcon::OnApplyTemplate()
+{
+    auto const grid = winrt::VisualTreeHelper::GetChild(*this, 0).as<winrt::Grid>();
```

Comment:
We use this pattern elsewhere:

    winrt::IControlProtected thisAsControlProtected = *this;

    winrt::ScrollPresenter scrollPresenter = GetTemplateChildT<winrt::ScrollPresenter>(s_scrollPresenterPartName, thisAsControlProtected);

Reply:
This doesn't work in this case because animated icon is not a normal templated control, we are using a "feature" of cppwinrt to inherit from IconElement through PathIcon. So we don't control our template, and this grid is created in IconElement's code behind and is not named.